app.py

```diff
@@ -41,152 +23,153 @@
+    # Parse out the start time, model number, model name, and status using regex
+    # Throw an error if the input doesn't look like dbt output
+    try:
+        df["post_regex"] = df["raw_line"].apply(
+            lambda col: re.search(REGEX, col).groups()
+        )
+    except AttributeError:
+        st.error(
+            "The input you provided doesn't look like dbt output. Please check your input and try again."
+        )
+        st.stop()
+
+    # Convert the post_regex column from a tuple to multiple columns
+    df[["start_time", "model_num", "model_name"]] = pd.DataFrame(
+        df["post_regex"].to_list(), index=df.index
+    )
+
+    # Convert columns to their correct data types
+    df["start_time"] = df.apply(
+        lambda row: str(datetime.strptime(row["start_time"], "%H:%M:%S").time()), axis=1
     )
-else:
-    st.subheader(f"Looks like there's {running_models} models still running. 🚀")
-    st.dataframe(df_running[["model_num", "start_time", "model_name"]])
-    st.caption(
-        "What do you think? Drop some feedback in [the repo](https://github.com/foundinblank/dbt-model-finder/) or email me at adamstone@gmail.com."
+    df["model_num"] = df["model_num"].astype(int)
+
+    # Get list of columns that are still running
+    # We are just looking for models that have a single line of output
+    still_running = (
+        df.groupby("model_num")["raw_line"]
+        .count()
+        .reset_index()
+        .rename(columns={"raw_line": "count_records"})
+        .query("count_records == 1")["model_num"]
+        .to_list()
     )
+
+    # Filter out the rows that are still running, clean up the columns and return
+    return (
+        df[df["model_num"].isin(still_running)]
+        .sort_values(by=["model_num"])
+        .drop(columns=["post_regex"])
+        .rename(
+            columns={
+                "model_num": "Model Number",
+                "start_time": "Start Time",
+                "model_name": "Model Name",
+            }
+        )
+        .reset_index(drop=True)
+    )
+
+
+def output(df):
+    '''
+    Output the results
+    '''
+    feedback_str = "What do you think? Drop some feedback in [the repo](https://github.com/foundinblank/dbt-model-finder/) or email me at adamstone@gmail.com."
+    running_models = df.shape[0]
+
+    if running_models == 0:
+        st.subheader("Hmm, I couldn't find any models still running! 🤔")
+    elif running_models == 1:
+        st.subheader("Looks like there's 1 model still running. 🤏")
+        st.dataframe(df[["Model Number", "Start Time", "Model Name"]])
+    else:
+        st.subheader(f"Looks like there's {running_models} models still running. 🚀")
+        st.dataframe(df[["Model Number", "Start Time", "Model Name"]])
+    st.caption(feedback_str)
+
+
+def main():
+    '''
+    Run the app
+    '''
+    title_and_description()
+    raw_input = get_logs()
+    df = clean_input(raw_input)
+    output(df)
```

Comment:
Learned a lot from this approach of wrapping up most things in functions, thank you for the demo!